modules/ItemLevel.lua

```diff
@@ -112,68 +112,64 @@ end
-				end
-				text:SetText(battlepetlevel)
-				return text:Show()
-			elseif text then
-				return text:Hide()
-			else
-				return
+	--Integration with SyLevel. If useSyLevel then let SyLevel handle item level logic.
+	if SyLevel then	
+		if settings.useSyLevel then
+			if text then
+				text:Hide()
 			end
+			SyLevel:CallFilters('Adibags', button, link)
+			return
+		else
+			SyLevel:CallFilters('Adibags', button, nil)
 		end
-
-		local _, _, quality, _, reqLevel, _, _, _, loc = GetItemInfo(link)
-		local item = Item:CreateFromBagAndSlot(button.bag, button.slot)
-		local level = item and item:GetCurrentItemLevel() or 0
-		if level >= settings.minLevel
-			and (quality ~= ITEM_QUALITY_POOR or not settings.ignoreJunk)
-			and (loc ~= "" or not settings.equippableOnly)
-			and (quality ~= ITEM_QUALITY_HEIRLOOM or not settings.ignoreHeirloom)
-		then
-			if SyLevel then
-				if settings.useSyLevel then
-					if text then
-						text:Hide()
-					end
-					SyLevel:CallFilters('Adibags', button, link)
-					return
-				else
-					SyLevel:CallFilters('Adibags', button, nil)
-				end
+	end
+	local level --The level to display for this item
+	local color --should be a table of color values to be passed to SetTextColor like returned by GetItemQualityColor()
+	local shouldShow = false --Set to true if this text should be shown
+	local link = button:GetItemLink()
+	--Item Logic
+	if link then
+		local objectTable = { strsplit(":", string.match(link, "|H(.-)|h")) } --see https://wow.gamepedia.com/ItemString for more info
+		if objectTable[1] == "item" then
+			local _, _, quality, _, reqLevel, _, _, _, loc = GetItemInfo(link)
+			local item = Item:CreateFromBagAndSlot(button.bag, button.slot)
+			level = item and item:GetCurrentItemLevel() or 0
+			if level >= settings.minLevel
+				and (quality ~= ITEM_QUALITY_POOR or not settings.ignoreJunk)
+				and (loc ~= "" or not settings.equippableOnly)
+				and (quality ~= ITEM_QUALITY_HEIRLOOM or not settings.ignoreHeirloom)
+			then
+				color = {colorSchemes[settings.colorScheme](level, quality, reqLevel, (loc ~= ""))}
+				shouldShow = true
 			end
-			if not text then
-				text = CreateText(button)
+		elseif objectTable[1] == "battlepet" then
+			if settings.showBattlePetLevels then
+				level = objectTable[3]
+				color = {GetItemQualityColor(objectTable[4])}
+				shouldShow = true
 			end
-			text:SetText(level)
-			text:SetTextColor(colorSchemes[settings.colorScheme](level, quality, reqLevel, (loc ~= "")))
-			return text:Show()
 		end
 	end
-	if SyLevel then
-		SyLevel:CallFilters('Adibags', button, nil)
-	end
-	if text then
-		text:Hide()
+	--Display Logic
+	if shouldShow then
+		if not text then
+			text = CreateText(button)
+		end
+		if level then
+			text:SetText(level)
+		end
+		if settings.colorScheme ~= "none" then
+			if color and #color >= 3 then
+				text:SetTextColor(unpack(color))
+			end
+		else
+			text:SetTextColor(colorSchemes["none"]())
+		end
+		text:Show()
```

Comment:
If we're optimizing here, are there cases where text would already be set where the level or color would change? So should some of this logic move inside the "if not text then" block? I suppose the settings could change between updates, so that would need to be accounted for.

And if there aren't cases where that level/color could change for an existing item level display, should the above block where we calculate level and color be skipped if text is present (and no settings change)?

Maybe overthinking... I'm not really sure how expensive any of these calls are.

Reply:
This plugin (ItemLevel.lua) creates fontstrings as needed when an AdiBags_UpdateButton event fires. In CreateText the fontstrings are stored in a table **texts[button]** where button is used as the key for the lookup.  Since mod:UpdateButton is triggered when AdiBags has a button update (new item, item moved, ect) The fonstrings are essentially tied to a slot and not to an item so it is possible and likely that when mod:UpdateButton is called the text and color have changed.

Reply:
This might be worth looking in to.  Equipping a new item triggers an AdiBags_UpdateButton event for visible button.  (in testing this was 28 times because I had 28 items in my bags) Caching could be done of the itemLink and if the itemLink hasn't changed then it doesnt need to do an update.